packages/component/src/SendBox/TypingIndicator.js

```diff
@@ -0,0 +1,47 @@
+import PropTypes from 'prop-types';
+import React, { useEffect, useState } from 'react';
+
+import { localize } from '../Localization/Localize';
+import connectToWebChat from '../connectToWebChat';
+import TypingAnimation from './Assets/TypingAnimation';
+
+const TypingIndicator = ({ language, lastTypingAt, styleSet }) => {
+  const {
+    options: { typingAnimationDuration }
+  } = styleSet;
+  const [showTyping, setShowTyping] = useState(false);
+
+  useEffect(() => {
+    let timeout;
+    const last = Math.max(Object.values(lastTypingAt));
+    if (last && Date.now() - last < typingAnimationDuration) {
```

Comment:
3 things.

## First

New line before `if`.

## Second

Also, you should calculate only once because:
- Save CPU
- The second time you do the calculation, it could be drifted, resulting in a negative value
   - `setTimeout` with a negative timeout value, in my experience, is bad. Maybe browsers are better handling them now

(Sorry I have a bad naming here, don't use that constant name if you can come up with a better one)

```js
if (last) {
  const typingAnimationDurationLeft = typingAnimationDuration - Date.now() + last;

  if (typingAnimationDurationLeft > 0) {
    setShowTyping(true);
    timeout = setTimeout(() => setShowTyping(false), typingAnimationDurationLeft);
  }
}
```

You can further reduce it into:

```js
const typingAnimationDurationLeft = typingAnimationDuration - Date.now() + last;

if (typingAnimationDurationLeft > 0) { // Always return false if last === undefined
  setShowTyping(true);
  timeout = setTimeout(() => setShowTyping(false), typingAnimationDurationLeft);
}
```

## Third

Corner cases. Think about, if the user want to have:

- Permanent typing animation
   - Intuitive, they will guess `Infinite` should do the trick
   - `true`, is kind of border case, I won't consider it given `Infinite` is so obvious
- Never show typing animation regardless
   - Intuitive, they will guess `0` or `false` should do the trick